router/router.go

```diff
@@ -0,0 +1,153 @@
+		dprequest.HandlerRequestID(16),
+		log.Middleware,
+		securityHandler,
+		healthcheckHandler(cfg.HealthCheckHandler),
+		serverError.Handler,
+		redirects.Handler,
+	}
+
+	alice := alice.New(middleware...).Then(router)
+
+	router.Handle("/", cfg.HomepageHandler)
+
+	router.Handle("/redir/{data:.*}", cfg.AnalyticsHandler)
+	router.Handle("/download/{uri:.*}", cfg.DownloadHandler)
+	router.Handle("/cookies{uri:.*}", cfg.CookieHandler)
+	router.Handle("/datasets/{uri:.*}", cfg.DatasetHandler)
+	router.Handle("/filters/{uri:.*}", cfg.FilterHandler)
+	router.Handle("/filter-outputs/{uri:.*}", cfg.FilterHandler)
+	router.Handle("/feedback{uri:.*}", cfg.FeedbackHandler)
+
+	if cfg.GeographyEnabled {
+		router.Handle("/geography{uri:.*}", cfg.GeographyHandler)
+	}
+
+	if cfg.SearchRoutesEnabled {
+		router.Handle("/search", cfg.SearchHandler)
+	}
+
+	// if the request is for a file go directly to babbage instead of using the allRoutesMiddleware
+	router.MatcherFunc(hasFileExtMatcher).Handler(cfg.BabbageHandler)
+
+	// If it is a known babbage endpoint go directly to babbage instead of using the allRoutesMiddleware
+	router.MatcherFunc(isKnownBabbageEndpointMatcher).Handler(cfg.BabbageHandler)
+
+	// all other requests go through the allRoutesMiddleware to check the page type first
+	allRoutesMiddleware := allRoutes.Handler(map[string]http.Handler{
+		"dataset_landing_page": cfg.DatasetHandler,
+	}, cfg.ZebedeeClient, cfg.ContentTypeByteLimit)
+
+	babbageRouter := router.PathPrefix("/").Subrouter()
+	babbageRouter.Use(allRoutesMiddleware)
+	babbageRouter.PathPrefix("/").Handler(cfg.BabbageHandler)
+
+	return alice
+}
+
+// securityHandler ...
+func securityHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.URL.Path != "/embed" && !strings.HasPrefix(req.URL.Path, "/visualisations/") {
+			w.Header().Set("X-Frame-Options", "SAMEORIGIN")
+		}
+		h.ServeHTTP(w, req)
+	})
+}
+
+// healthcheckHandler uses the provided handler for /health endpoint, and serves any other traffic to the next handler in chain
+func healthcheckHandler(hc func(w http.ResponseWriter, req *http.Request)) func(h http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			if req.URL.Path == "/health" {
+				hc(w, req)
+				return
+			}
+			h.ServeHTTP(w, req)
+		})
+	}
+}
+
+var knownBabbageEndpoints = []string{
+	"/chartconfig",
+	"/chartimage",
+	"/generator",
+	"/visualisations/",
+	"/timeseriestool",
+	"/search",
+	"/file",
+	"/resource",
+	"/ons/",
+	"/file",
+	"/calendar",
+	"/chart",
+	"/embed",
+	"/export",
+	"/hash",
+}
+
+// IsKnownBabbageEndpoint returns true if the given path matches a known babbage endpoint
+func IsKnownBabbageEndpoint(path string) bool {
+	for _, endpoint := range knownBabbageEndpoints {
+		if strings.HasPrefix(path, endpoint) {
+			return true
+		}
+	}
+	return false
```

Comment:
Logically this seems fine but I am slightly hesitant that these endpoint values alone aren't specific enough to be able to make this determination conclusively.  If we ever had 2 endpoints named similarly in 2 different APIs this _might_ lead to incorrect routing and some weird issues.
This is probably me being overly cautious and I don't want to create an unnecessary headache for you so as a compromise could we add a log event to say something like `identified babbage endpoint redirecting...`? Or something like that - it's a breadcrumb for us to follow if this potential  issue ever became a problem and give us a clue what was happening.

Reply:
A good point, though this specific Babbage related matching is only used as a last resort when a request is not matched to any other API. If a new endpoint was added to an API it would have to be added above the Babbage fallback routing.

I was about to add the logging as suggested, but looking at the existing logs I think there is enough to suggest where a request is being routed to. If a new endpoint was added and not being routed correctly the logs could be searched for the URL to identify the proxy being used. Here is an trimmed down example:

```
{
  "data": {
    "proxy_name": "babbage"
  },
  "event": "proxying request",
  "http": {
    "method": "GET",
    "path": "/generator",
    "query": "uri=/economy/environmentalaccounts/bulletins/ukenvironmentalaccounts/2015-07-09/47863209&format=csv&"
  },
  "namespace": "dp-frontend-router",
}
```

Reply:
I forgot to mention - I made the matching more specific for the known paths. This means if another API used a path with one of the known endpoints as a prefix, it would not be identified as a Babbage endpoint. (even though it should be matched before reaching the Babbage related routes)